upcloud/resource_upcloud_server.go

```diff
@@ -516,155 +562,69 @@ func buildServerOpts(d *schema.ResourceData, meta interface{}) (*request.CreateS
-	client := meta.(*service.Service)
-
-	storageDevices := d.Get("storage_devices").([]interface{})
-
-	r := &request.GetServerDetailsRequest{
-		UUID: d.Id(),
-	}
-	server, err := client.GetServerDetails(r)
-	if err != nil {
-		return err
-	}
-
-	for i, storageDevice := range storageDevices {
-		storageDevice := storageDevice.(map[string]interface{})
-
-		if backupRule := storageDevice["backup_rule"].(*schema.Set).List(); backupRule != nil && len(backupRule) != 0 {
-
-			for _, br := range backupRule {
-				mBr := br.(map[string]interface{})
-
-				retentionValue, err := strconv.Atoi(mBr["retention"].(string))
+	if template, ok := d.GetOk("template.0"); ok {
+		template := template.(map[string]interface{})
+		if template["title"].(string) == "" {
+			template["title"] = fmt.Sprintf("terraform-%s-disk", r.Hostname)
+		}
+		serverStorageDevice := request.CreateServerStorageDevice{
+			Action:  "clone",
+			Address: template["address"].(string),
+			Size:    template["size"].(int),
+			Storage: template["storage"].(string),
+			Title:   template["title"].(string),
+		}
+		if attr, ok := d.GetOk("template.0.backup_rule.0"); ok {
+			serverStorageDevice.BackupRule = backupRule(attr.(map[string]interface{}))
+		}
+		if source := template["storage"].(string); source != "" {
+			// Assume template name is given and attempt map name to UUID
+			if _, err := uuid.ParseUUID(source); err != nil {
+				l, err := meta.(*service.Service).GetStorages(
+					&request.GetStoragesRequest{
+						Type: upcloud.StorageTypeTemplate,
+					})
 
 				if err != nil {
-					diag.FromErr(err)
-				}
-
-				modifyStorage := &request.ModifyStorageRequest{
-					UUID: server.StorageDevices[i].UUID,
+					return nil, err
 				}
-
-				modifyStorage.BackupRule = &upcloud.BackupRule{
-					Interval:  mBr["interval"].(string),
-					Time:      mBr["time"].(string),
-					Retention: retentionValue,
+				for _, s := range l.Storages {
+					if s.Title == source {
+						source = s.UUID
+						break
+					}
 				}
-
-				client.ModifyStorage(modifyStorage)
 			}
-		}
-	}
-
-	return nil
-}
-
-func buildStorage(storageDevice map[string]interface{}, i int, meta interface{}, hostname, zone string) (*request.CreateServerStorageDevice, error) {
-	osDisk := request.CreateServerStorageDevice{}
 
-	if source := storageDevice["storage"].(string); source != "" {
-		_, err := uuid.ParseUUID(source)
-		// Assume template name is given and map name to UUID
-		if err != nil {
-			client := meta.(*service.Service)
-			r := &request.GetStoragesRequest{
-				Type: upcloud.StorageTypeTemplate,
-			}
-			l, err := client.GetStorages(r)
-			if err != nil {
-				return nil, err
-			}
-			for _, s := range l.Storages {
-				if s.Title == source {
-					source = s.UUID
-					break
-				}
-			}
+			serverStorageDevice.Storage = source
+		}
+		r.StorageDevices = append(r.StorageDevices, serverStorageDevice)
+		// TODO: handle backup_rule
```

Comment:
Was this forgotten or just documented for future work?

Reply:
was forgotten. good catch!